feat: adds "data collection for marketing" toggles

jonybur opened the pull request:
Adds data collection for marketing toggles (and toasts/warnings) on:
- Onboarding
- Toast in Wallet
- Settings page

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:
https://github.com/MetaMask/MetaMask-planning/issues/2525
https://github.com/MetaMask/MetaMask-planning/issues/2439
https://github.com/MetaMask/MetaMask-planning/issues/2440

## **Manual testing steps**


Onboarding checkbox:
Set `isPastPrivacyPolicyDate` to true in OptInMetrics
1. Start a new account
2. There should be a new checkbox that asks for marketing consent
3. Checking it should set the marketing consent to true (check at Settings, Securty tab page)

Security tab:
1. Go to Security tab
2. When checking the "Data collection for marketing" to `true`, the "Participate in MetaMetrics" toggle should turn to `true`
3. When checking "Participate in MetaMetrics" to `false`, "Data collection for marketing" should be set to `false`
4. When "Participate in Metametrics" is `true` and "Data collection for marketing" is `true`, and the latter is set to `false`, a warning message should appear.

Toast:
An already onboarded user will have the "dataCollectionForMarketing" value as `null` (neither `true` or `false`). This will trigger the toast.
1. By clicking on "I accept", it should set the "Data collection for marketing" to `true`.
2. By closing the toast or clicking on "No thanks", it should set the "Data collection for marketing" to `false`.

All of these actions should trigger subsequent Segment events.


## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**


https://github.com/MetaMask/metamask-mobile/assets/11148144/d0ebee90-38e1-4362-92ad-73728c82df72

<img width="573" alt="Screenshot 2024-05-20 at 19 18 19" src="https://github.com/MetaMask/metamask-mobile/assets/11148144/04538fdb-564e-4c75-9922-f0224bcbeec6">

<img width="573" alt="Screenshot 2024-05-20 at 19 19 30" src="https://github.com/MetaMask/metamask-mobile/assets/11148144/347bf76a-e075-4647-b015-b1770814556b">



## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Commit:
Add checkbox to onboarding

Commit:
Add copy

Commit:
Add BottomSheet

Commit:
Add checkbox

Commit:
Add data collection modal

Commit:
Merge branch 'jb-metametrics-onboarding-checkbox' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Merge branch 'jb-metametrics-id-toast' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Merge branch 'jb-metametrics-id-settings' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Add redux support

Commit:
Use message box for Data Collection disable warning

Commit:
Connect DataCollectionModal

Commit:
Add tracking events

Commit:
Merge branch 'main' into jb-opt-in-metrics

EtherWizard33 reviewed (commented):
as an existing wallet, a bottom sheet show up, when I click 'Agree', and go to the marketing setting the switch is off an expected to be on since 'Agree' was clicked.


<img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/MetaMask/metamask-mobile/assets/165834542/d1c692e2-0ae9-45c8-bfc5-e6dd92a992ed">

EtherWizard33 reviewed (commented):
I already discussed it with Jony, about the scenario showing one metrics screen + later a bottom sheet both in the onboarding flow, it is that that it is because it's the legacy onboarding, I still wanted to keep a recording of it incase this scenario needs to be viewed/reviewed
<div>
    <a href="https://www.loom.com/share/edd5ae375b8443de9c526592cd6a81c2">
      <p>Simulator - iPad (10th generation) - 21 May 2024 - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/edd5ae375b8443de9c526592cd6a81c2">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/edd5ae375b8443de9c526592cd6a81c2-with-play.gif">
    </a>
  </div>

jonybur commented:
@eriknson Just confirmed that this will show after the new privacy policy is rolled out, so the legacy onboarding scenario won't be an issue

Commit:
Fix dispatch

jonybur commented:
@EtherWizard33 The issue with the toast should be fixed now

Commit:
Merge branch 'jb-opt-in-metrics' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Merge branch 'main' into jb-opt-in-metrics

Commit:
Only show toast if participating in metametrics

Commit:
Merge branch 'main' into jb-opt-in-metrics

Commit:
Add missing links

Commit:
Merge branch 'jb-opt-in-metrics' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Implement Add traits to user

Commit:
Update snapshot

Commit:
Add dataCollectionForMarketing on mockstate

Commit:
Merge branch 'main' into jb-opt-in-metrics

Commit:
Change test id

EtherWizard33 reviewed (dismissed):
LGTM

darkwing commented on app/components/UI/OptinMetrics/index.js:
```diff
@@ -337,6 +348,21 @@ class OptinMetrics extends PureComponent {
+        metrics.trackEvent(MetaMetricsEvents.ANALYTICS_PREFERENCE_SELECTED, {
+          ...traits,
+          location: 'onboarding_metametrics',
+        });
```
Nit: Does `traits` need to be added via `addTraitsToUser` *and* on line 361?

jonybur replied:
Yes, these are two different methods. The `addTraitsToUser` affects the segment configuration, and the `trackEvent` pings Segment with the tracking event.

darkwing reviewed (dismissed):
Looks good to me, just one nit!

Commit:
Update snapshot

Commit:
Add snapshot

Commit:
Add snapshot

Commit:
Add reducer

Commit:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

EtherWizard33 committed:
test: added test on of content rendering

EtherWizard33 committed:
test: update snap and lint

sethkfman committed:
chore: remove enzyme (#9741)

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

EtherWizard33 committed:
test: add e2e checking the enhance experience bottom sheet is showing when expected, dismiss and confirm it's dimissed

EtherWizard33 committed:
style: lint

Commit:
Add tests

Commit:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

tommasini commented on app/actions/security/state.ts:
```diff
@@ -3,4 +3,5 @@ export interface SecuritySettingsState {
   automaticSecurityChecksEnabled: boolean;
   hasUserSelectedAutomaticSecurityCheckOption: boolean;
   isAutomaticSecurityChecksModalOpen: boolean;
+  dataCollectionForMarketing: boolean | null;
```
I think if we do not want to follow the pattern of creating another variable to tell us if the collection modal allowance was already answered, we should document properly that the null represents the modal not being shown to the end user yet! Although I would say I prefer to separate both responsibilities splitter and clear on the codebase

jonybur replied:
Added a comment

tommasini reviewed (changes requested):
Heyy, good work! Left some comments

jonybur commented on app/reducers/security/index.ts:
```diff
@@ -7,6 +7,7 @@ export const initialState: Readonly<SecuritySettingsState> = {
   automaticSecurityChecksEnabled: false,
   hasUserSelectedAutomaticSecurityCheckOption: false,
   isAutomaticSecurityChecksModalOpen: false,
+  dataCollectionForMarketing: null,
```
Not really. `null` is if the value has not been set yet, which is different from having it to `false`.

Commit:
Add comment

EtherWizard33 committed:
Revert "test: add e2e checking the enhance experience bottom sheet is showing when expected, dismiss and confirm it's dimissed"

This reverts commit d2186d1eb80f01d6e59dab7ea5ba7c46f40b3fde.

Commit:
Add feature flag to opt in bottom sheet

Commit:
Merge branch 'jb-opt-in-metrics' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Fix var

cortisiko committed:
Merge branch 'main' into jb-opt-in-metrics

tommasini commented on app/components/Views/DataCollectionModal/index.tsx:
```diff
@@ -0,0 +1,63 @@
+    onPress: () => {
+      bottomSheetRef.current?.onCloseBottomSheet();
+    },
+    testID: 'accept-button',
```
Nit: Can we move this to a constante and rename it so we know this is the accept button of data collection modal component?

Something like `collection-modal-accept-button`? Wdyt?

EtherWizard33 replied:
done, marking as resolved

tommasini commented on app/components/Views/DataCollectionModal/index.tsx:
```diff
@@ -0,0 +1,63 @@
+          size={IconSize.Lg}
+          name={IconName.Warning}
+          color={IconColor.Warning}
+          testID="icon-warning"
```
Nit: Can we move the test id to a constant?

EtherWizard33 replied:
done, marking as resolve

tommasini commented on app/components/Views/ExperienceEnhancerModal/index.tsx:
```diff
@@ -0,0 +1,118 @@
+        location: 'marketing_consent_modal',
+      });
+    },
+    testID: 'cancel-button',
```
Nit: Can we move the test id to a constant?

EtherWizard33 replied:
done, marking as resolve

tommasini commented on app/components/Views/ExperienceEnhancerModal/index.tsx:
```diff
@@ -0,0 +1,118 @@
+            variant={ButtonVariants.Link}
+            label={strings('experience_enhancer_modal.link')}
+            onPress={() => Linking.openURL(HOW_TO_MANAGE_METRAMETRICS_SETTINGS)}
+            testID="link-button"
```
Nit: Can we move the test id to a constant?

EtherWizard33 replied:
done, marking as resolve

tommasini commented on app/components/Views/ExperienceEnhancerModal/index.tsx:
```diff
@@ -0,0 +1,118 @@
+        location: 'marketing_consent_modal',
+      });
+    },
+    testID: 'accept-button',
```
Nit: Can we move the test id to a constant?

EtherWizard33 replied:
done, marking as resolved

tommasini commented on app/components/Views/ExperienceEnhancerModal/index.tsx:
```diff
@@ -0,0 +1,118 @@
+  };
+
+  return (
+    <BottomSheet ref={bottomSheetRef} testID="bottom-sheet">
```
Nit: Can we move the test id to a constant?

EtherWizard33 replied:
done, marking as resolve

tommasini replied:
This still appears to me as hardcoded

tommasini commented on app/components/Views/ExperienceEnhancerModal/index.tsx:
```diff
@@ -0,0 +1,118 @@
+          {strings('experience_enhancer_modal.paragraph2')}
+        </Text>
+        <View style={styles.list}>
+          <Text style={styles.line} testID="bullet1">
```
Nit: Can we move the test id to a constant?

EtherWizard33 replied:
done, marking as resolve

tommasini commented on app/components/Views/ExperienceEnhancerModal/index.test.tsx:
```diff
@@ -0,0 +1,83 @@
+  it('should handle cancel button press correctly', () => {
+    const { getByTestId } = render(<ExperienceEnhancerModal />);
+
+    const cancelButton = getByTestId('cancel-button');
```
Nit: Can we add the constant created here?

EtherWizard33 replied:
done, marking as resolve

tommasini commented on app/components/Views/Settings/SecuritySettings/SecuritySettings.tsx:
```diff
@@ -468,11 +476,47 @@ const Settings: React.FC = () => {
     } else {
       await enable(false);
       setAnalyticsEnabled(false);
+      if (isDataCollectionForMarketingEnabled) {
```
Just to confirm, if the `isDataCollectionForMarketingEnabled` value is null, we do not want it to be set to false? 
If it's not possible for the variable to be null at this point, would we need the conditition to check if its true?

EtherWizard33 replied:
done, marking as resolve

tommasini commented on app/components/Views/Wallet/index.tsx:
```diff
@@ -170,6 +171,10 @@ const Wallet = ({
   const prevChainId = usePrevious(providerConfig.chainId);
+
+  const isDataCollectionForMarketingEnabled = useSelector(
+    (state: any) => state.security.dataCollectionForMarketing,
```
We shouldn't use any, if possible we should create create the type, or we can say that its an expected error if it will be address soon

Any remove any type safet that we get from using typescript

I believe we have a RootState with the types defined, could you update it to have this new variable? 🙏🏼

EtherWizard33 replied:
done, marking as resolve

tommasini reviewed (changes requested):
Everything looks good, but spotted an "any" insertion, and we should avoid it at all cost.

Left also some nit picks coments and a comment with a question

EtherWizard33 committed:
style: ts types

EtherWizard33 committed:
refactor: extract ExperienceEnhancerModal constants

EtherWizard33 committed:
refactor: extract DataCollectionModal constants

EtherWizard33 committed:
feat: replace loose equality by strict equality

EtherWizard33 committed:
refactor: extract DataCollectionModal constants

tommasini reviewed (dismissed):
LGTM! Just seem that two hardcoded test ids were missing

EtherWizard33 committed:
refactor: extract ExperienceEnhancerModal constants

cortisiko commented on app/components/Views/DataCollectionModal/index.constants.ts:
```diff
@@ -0,0 +1,4 @@
+export default {
```
can the test IDs go into the e2e/selectors folder for this particular modal? this keeps all the testIDs in a centralized location

EtherWizard33 replied:
Hey Curtis, yes you are right, this is now done, marking as resolved.

cortisiko commented on app/components/Views/ExperienceEnhancerModal/index.constants.ts:
```diff
@@ -0,0 +1,13 @@
+export default {
```
can the test IDs go into the e2e/selectors folder for this particular modal? this keeps all the testIDs in a centralized location

EtherWizard33 replied:
Hey Curtis, yes you are right, this is now done, marking as resolved.

cortisiko reviewed (dismissed):
Small feedback @EtherWizard33

EtherWizard33 committed:
refactor: move test id constants to e2e selectors

EtherWizard33 committed:
feat: change the privacy policy date from June 4th to June 18th as per Hester

Commit:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

tommasini commented:
Can we update the initial-background-state and fixture-builder files to have this new data?

Commit:
Add dataCollectionForMarketing to fixtureBuilder

tommasini commented on app/components/Views/Settings/SecuritySettings/SecuritySettings.tsx:
```diff
@@ -468,11 +476,47 @@ const Settings: React.FC = () => {
     } else {
       await enable(false);
       setAnalyticsEnabled(false);
+      if (isDataCollectionForMarketingEnabled === true) {
```
```suggestion
      if (isDataCollectionForMarketingEnabled) {
```

Commit:
Update app/components/Views/Settings/SecuritySettings/SecuritySettings.tsx

Co-authored-by: tommasini <46944231+tommasini@users.noreply.github.com>

Commit:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Merge branch 'jb-opt-in-metrics' of github.com:MetaMask/metamask-mobile into jb-opt-in-metrics

Commit:
Revert fb change

tommasini reviewed (approved):
LGTM!

salimtb committed:
Merge branch 'main' into jb-opt-in-metrics

hesterbruikman commented:
Removed the auto-added 7.25.0 label as this is planned for a hotfix based off of 7.24.0 (more on Slack). cc @jonybur @darkwing @Cal-L @NicolasMassart
